calico.go

```diff
@@ -0,0 +1,302 @@
+	// profile creation.
+	if !runningUnderK8s || conf.Policy == nil {
+		//TODO - this is the wrong test. It should be on the policy type
+		// Start by checking if the profile already exists. If it already exists then there is no work to do (the CNI plugin never updates a profile).
+		exists, err := libcalico.ProfileExists(conf.Name, etcd)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			// The profile doesn't exist so needs to be created. The rules vary depending on whether k8s is being used.
+			// Under k8s (without full policy support) the rule is permissive and allows all traffic.
+			// Otherwise, incoming traffic is only allowed from profiles with the same tag.
+			k8sInboundRule := []libcalico.Rule{libcalico.Rule{Action: "allow"}}
+			tagInboundRule := []libcalico.Rule{libcalico.Rule{Action: "allow", SrcTag: conf.Name}}
+			fmt.Fprintf(os.Stderr, "Calico CNI creating profile. profile=%s\n", conf.Name)
+
+			var inboundRule []libcalico.Rule
+			if runningUnderK8s {
+				inboundRule = k8sInboundRule
+			} else {
+				inboundRule = tagInboundRule
+			}
+
+			profile := libcalico.Profile{
+				ID: conf.Name,
+				Rules: libcalico.Rules{
+					Inbound:  inboundRule,
+					Outbound: []libcalico.Rule{libcalico.Rule{Action: "allow"}}},
+				Tags: []string{conf.Name}}
+			if err := profile.Write(etcd); err != nil {
+				return err
+			}
+		}
+	}
+
+	return result.Print()
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	n := NetConf{}
+	if err := json.Unmarshal(args.StdinData, &n); err != nil {
+		return fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	if err := AddIgnoreUnknownArgs(); err != nil {
+		return err
+	}
+
+	// Always try to release the address
+	if err := ipam.ExecDel(n.IPAM.Type, args.StdinData); err != nil {
+		return err
+	}
+
+	// Always try to clean up the workload/endpoint.
+	// First determine if running under k8s to get the right workload and orchestrator IDs
+	k8sArgs := K8sArgs{}
+	if args.Args != "" {
+		err := LoadArgs(args.Args, &k8sArgs)
+		if err != nil {
+			return err
+		}
+	}
+	var orchestratorID, workloadID string
+
+	runningUnderK8s := string(k8sArgs.K8S_POD_NAMESPACE) != "" && string(k8sArgs.K8S_POD_NAME) != ""
+	if runningUnderK8s {
+		workloadID = fmt.Sprintf("%s.%s", k8sArgs.K8S_POD_NAMESPACE, k8sArgs.K8S_POD_NAME)
+		orchestratorID = "k8s"
+	} else {
+		workloadID = args.ContainerID
+		orchestratorID = "cni"
+	}
+
+	// Actually remove the workload
+	etcd, err := libcalico.GetKeysAPI(n.EtcdAuthority, n.EtcdEndpoints)
+	if err != nil {
+		return err
+	}
+	workload := libcalico.Workload{
+		Hostname:       hostname,
+		OrchestratorID: orchestratorID,
+		WorkloadID:     workloadID}
+	if err := workload.Delete(etcd); err != nil {
+		return err
+	}
+
+	// Only try to delete the device if a namespace was passed in
+	if args.Netns != "" {
+		err := ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+			var err error
```

Comment:
Don't think you need to declare the variable `err` right here, unless I'm missing something.